### PR TITLE
This adds the summaries to the tgene artifacts

### PR DESCRIPTION
**Pull request by antgonza**
Based on https://github.com/biocore/qiita/pull/1697 so review that one first. 

The summaries being created are like:
screen shot 2016-03-19 at 7 18 12 pm

### CODE REVIEW

**Comment by antgonza** [comment]
After discussing with @josenavas we agreed on also adding the demux summary here. The html is going to look like:
screen shot 2016-03-20 at 8 23 02 am

**Comment by antgonza** [comment]
Ready for review. Note that this PR also updating the branch to master so ignore changes to:
`qiita_plugins/target_gene/tgp/split_libraries/tests/test_util.py`
`qiita_plugins/target_gene/tgp/split_libraries/util.py`

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+from qiita_ware.demux import stats as demux_stats
+
+
+FILEPATH_TYPE_TO_NOT_SHOW_HEAD = ['SFF']
+LINES_TO_READ_FOR_HEAD = 10
+
+
+def generate_html_summary(qclient, job_id, parameters, out_dir,
+                          return_html=False):
+    """Generates the HTML summary of a BIOM artifact
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: `BIOM artifact` -> `target gene type artifact`

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    Returns
+    -------
+    dict
+        The results of the job
+
+    Raises
+    ------
+    ValueError
+        - If there is any error gathering the information from the server
+        - If there artifact is 'Demultiplexed' but it doesn't have a demux file
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: `If there artifact is` -> `If the artifact is of type`

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        artifact_information = _summary_demultiplexed(
+            artifact_type, fps_info['filepaths'])
+        if artifact_information is None:
+            raise ValueError("We couldn't find a demux file in your artifact")
+    else:
+        artifact_information = _summary_not_demultiplexed(
+            artifact_type, fps_info['filepaths'])
+
+    of_fp = join(out_dir, "artifact_%d.html" % artifact_id)
+    of = open(of_fp, 'w')
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]:
```python
with open(of_fp, 'w') as of:
    of.write('\n'.join(artifact_information))
```

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        if artifact_type not in FILEPATH_TYPE_TO_NOT_SHOW_HEAD:
+            # we need to encapsulate the full for loop because gzip will not
+            # raise an error until you try to read
+            try:
+                with gopen(fps, 'r') as fin:
+                    for i, line in enumerate(fin):
+                        header.append(line)
+                        if i >= LINES_TO_READ_FOR_HEAD:
+                            break
+            except:
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: Can you specify the error that is actually being raised?

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                hash_md5.update(chunk)
+
+        # getting head of the files
+        header = []
+        if artifact_type not in FILEPATH_TYPE_TO_NOT_SHOW_HEAD:
+            # we need to encapsulate the full for loop because gzip will not
+            # raise an error until you try to read
+            try:
+                with gopen(fps, 'r') as fin:
+                    for i, line in enumerate(fin):
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: Can you use the first example in [this link](http://stackoverflow.com/a/1767589/3746629) to read the first X lines of the file? Use also in line 122.

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    of.write('\n'.join(artifact_information))
+
+    # Step 3: add the new file to the artifact using REST api
+    reply = qclient.patch(qclient_url, 'add', '/html_summary/',
+                          value=of_fp)
+
+    return reply if not return_html else (reply, artifact_information)
+
+
+def _summary_not_demultiplexed(artifact_type, filepaths):
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: Can you add docs?

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                                    hash_md5.hexdigest())
+        if header:
+            artifact_information.append(
+                "<p style=\"font-family:'Courier New', Courier, monospace;"
+                "font-size:10;\">%s</p><hr/>" % ("<br/>".join(header)))
+
+    return artifact_information
+
+
+def _summary_demultiplexed(artifact_type, filepaths):
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: Can you add docs?

File: qiita_plugins/target_gene_type/target_gene_type/summary.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+import matplotlib.pyplot as plt
```
Comment (josenavas) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: can you put this import below the one from `qiita_ware.demux`? (pep8 import order)

File: qiita_plugins/target_gene_type/target_gene_type/summary.py (hunk 4 above)
Comment (ElDeveloper) [inline qiita_plugins/target_gene_type/target_gene_type/summary.py]: :+1:

**Comment by ElDeveloper** [comment]
:+1: looks good.

**Comment by ElDeveloper** [comment]
Just noticed, the plot seems rather odd, but I assume this is because we are using "small test data"? How does this look with regular datasets?

**Comment by antgonza** [comment]
This is not a large dataset but a bit larger
screen shot 2016-03-22 at 1 50 55 pm

**Comment by ElDeveloper** [comment]
Thanks!

**Comment by ElDeveloper** [comment]
For the curious:

screen shot 2016-03-22 at 1 59 25 pm

**Pull request merged**